### ISSUE TITLE
Installation Guide

### ISSUE THREAD

**Issue by kaznad** [albertjan/PipeToCom#10]
Hi, Please advise how do I install this service? Thanks :)

**Comment by PiterEL** [comment]
Hello
I have a problem.
I run this OK:
%WINDIR%\Microsoft.NET\Framework\v4.0.30319\InstallUtil -t Path\To\NP2COMS.exe

Then i run NP2COMV to generate new config file. I setup parameters and try to save. I have an error on the screen.
![bug](https://cloud.githubusercontent.com/assets/4134018/6436634/78274192-c0b3-11e4-8da5-9af6443828ae.jpg)

I have installed Microsoft .NET Framework 4.5.1 (4.5.50938)

Please help me...

**Comment by eddelac** [comment]
I've followed the install steps with success, registered the service, created the named pipe, created the config file using the named pipe, saved the config file on the same path, deleted the example config.

But the service stops just after starting it. The virtual machine is on. I'm im missing something?

**Comment by dcrowson** [comment]
I found this to happen if the device using the com/serial port wasn’t running or connected.

 

Dan

----

Dan Crowson

IS Director

CMS Communications, Inc.

722 Goddard Ave

Chesterfield MO 63005

dcrowson@cmsc.com <mailto:dcrowson@cmsc.com>  

www.cmsc.com <http://www.cmsc.com/>  

Expertise you need, the choice you deserve.

**Comment by eddelac** [comment]
@PiterEL I had the same problem, it seems to be related with the folder permissions. Try running the program as Administrator.

**Comment by albertjan** [comment]
@PiterEL It might be due to the fact your running vmware. It creates a namedpipe with a weird name. @eddelac Might be right aswel. 

@PiterEL the config files are really easy to edit by hand. have a look at `example.n2c` you can probably edit that to fit your needs.

**Comment by albertjan** [comment]
@eddelac did you fix your problem? There is a log file in the directory. If you could open a new issue and post that logfile there. I can have a look and see whats going wrong.

**Comment by fgmike** [comment]
Hi,

Getting the same error as @PiterEL. The named pipe in the example file, should I change that from com to \\.\pipe\comport for example?

How else should I edit it, as I cannot see an existing named pipe on my Hyper-V Server 2012 R2 listed when I run np2comv.

Any chance also of fixing the image examples in your setup instructions above? I could not see it to compare.

Trying to get my head around these named pipes is causing some extra grey hairs.

Any help Appreciated.

**Comment by dayers** [comment]
Hi,

Just set this up successfully mapping a hyperv vm com port to a physical com port.  I followed the steps below in order:

1. Install tera term on host machine to give you something to test your terminal connection with.
2. Install driver for usb to serial cable if required.
3. Insert usb to serial cable.
4. Launch tera term - select serial - check com usb to serial cable inserted in is port listed. Use device manager com ports or at a command prompt type c:\change port /query.
5. Install NP2com -   You can run %WINDIR%\Microsoft.NET\Framework\v4.0.30319\InstallUtil -t Path\To\NP2COMS.exe.
6. Create named pipe for vm - launch hyperV manager - highlight VM (vm off at this point) - select settings in lower windows - com1 - enter a name for pipe e.g comportvm.  Leave remote computer box inactive.
7. Start VM (login not reuired at this point).
8. Run NP2com\NP2comv.exe to create a N2C profile;
     Select pipe name from right hand drop down box - comportvm - this is near the bottom of the list
      Set the baud rate, stop bits and parity i.e 9600, 8, none, 1

**Comment by vibac** [comment]
albertjan,

I have a questions regarding the install and I am hoping you or someone here can assist me, do we need to complete this installation on both host and guest?

I also keep getting this error message when trying to register the service; 
C:\Windows\system32>c:\windows\microsoft.net\framework\v4.0.30319\installutil.ex
e C:\NP2COM\NP2COMS.exe
Microsoft (R) .NET Framework Installation utility Version 4.0.30319.17929
Copyright (C) Microsoft Corporation.  All rights reserved.

Exception occurred while initializing the installation:
System.IO.FileLoadException: Could not load file or assembly 'file:///C:\NP2COM\
NP2COMS.exe' or one of its dependencies. Operation is not supported. (Exception
from HRESULT: 0x80131515).

**Comment by albertjan** [comment]
Do you have .net 3.5 and less installed? ;)

**Comment by vibac** [comment]
yes 4.5, I was able to install and start the NP2COMS service successfully after I posted my questions. ;)

I was asking because I couldn't get the service registered but found that I had to unblock it. Please bare with me as I am a system admin and have very little programming knowledge.

BTW, thank you very much for the quick reply and for your work, we greatly appreciate.

Once I have it installed and the service started can I just go ahead and start using the guest's COM port or do I also need to install it on the guest?

**Comment by albertjan** [comment]
If you hooked it up to the correct namedpipe you should be fine.

**Comment by vibac** [comment]
It is, I will only now once I have finished installing our soft on the guest. I will let you know, thx a million sir ;)

**Comment by PiterEL** [comment]
Hello
I want use faxmodem Dell connected to physical serwer in serwer Hyper-V win2008 R2. Modem is connected on USB and it is show in physical machine as modem on COM3.
I do all step by step as @dayers describe and i run Hyper-V machine. How i can test and add Faxmodem to machine Hyper-V ? Please help me...

On physical machine service NP2COM works ok. Please help me because i want run service serwer fax on my virtual machine Hyper-V.

**Comment by vibac** [comment]
Hello albertjan,

All is good!!! Thanks :)

I will post all my steps, please let me know that's ok with you.

=========================
--How to Install NP2COM--
=========================
Give your user full rights on the NP2COM folder after extracting it
1- Right click the NP2COM folder>properties and give your user full rights
1a- You can put this folder anywhere on your server, ie: C:\NP2COM\

Ensure that the NP2COMS general properties are unblocked
2- Right click the NP2COMS.exe>properties at the bottom hit the "unblock" button

Register the service NP2COMS
3- After you've installed .net 3, 3.5, 4 or 4.5 you can register NP2COMS.exe as a service by applying the following cmd

Start cmd with "run as Admin"
c:\windows\microsoft.net\framework\v4.0.30319\installutil.exe -t C:\NP2COM\NP2COMS.exe
or use run
%WINDIR%\Microsoft.NET\Framework\v4.0.30319\InstallUtil -t C:\NP2COM\NP2COMS.exe

Create the Named pipe in the guest VM settings
4- Right click the vm from within the host's Hyper-V, select "settings", then select com1 or com2, tic "named pipe" and enter a name, ie: comport

Create a config file for NP2COMs service
5- Run the NP2COMV.exe as admin to create a .n2c config file and save it in the Path\To\NP2COMS.exe, ie: C:\NP2COM\
5a- On the left hand side select the NamedPipe you created previously in step 4 from the drop down list, ie: \\.\pipe\comport
5b- Set the serial port settings to your needs, ie: baudrate 9600
5c- Now click "Write config file" to save it and save it in the NP2COM location, ie: C:\NP2COM\

Delete the example config file
6- Delete example.n2c

Start the NP2COMS service
7- Run: c:\>net start NP2COMService or go to services and start it

DONE!

Test communication
1- Stop the NP2COM service on host
2- Launch the NP2COMV.exe as admin
3- Go to the guest and launch hyper terminal or putty terminal
4- Once you connect to the setup COM port hit enter
5- Go to the NP2COMV window on the host and see if the data is written
6- Once you are done, kill the NP2COMV process then start the NP2COM service

**Comment by albertjan** [comment]
@vibac thanks! :+1:

**Comment by sonvoba** [comment]
I got the following unhandled exception error when I clicked on "Write config file" button of NP2COMV.  Screen Shot and Details are attached.

![untitled 06e](https://cloud.githubusercontent.com/assets/10577963/9076515/e7e67b1c-3ad7-11e5-9934-90a78b4a20ae.png)

See the end of this message for details on invoking 
just-in-time (JIT) debugging instead of this dialog box.

************** Exception Text **************
System.ArgumentNullException: Value cannot be null.
Parameter name: input
   at System.Text.RegularExpressions.Regex.Match(String input)
   at System.Text.RegularExpressions.Regex.Match(String input, String pattern)
   at NP2COMV.Form1.button2_Click(Object sender, EventArgs e)
   at System.Windows.Forms.Control.OnClick(EventArgs e)
   at System.Windows.Forms.Button.OnClick(EventArgs e)
   at System.Windows.Forms.Button.OnMouseUp(MouseEventArgs mevent)
   at System.Windows.Forms.Control.WmMouseUp(Message& m, MouseButtons button, Int32 clicks)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ButtonBase.WndProc(Message& m)
   at System.Windows.Forms.Button.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)


************** Loaded Assemblies **************
mscorlib
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.79.0 built by: NETFXREL2
    CodeBase: file:///C:/Windows/Microsoft.NET/Framework64/v4.0.30319/mscorlib.dll
----------------------------------------
NP2COMV
    Assembly Version: 1.0.0.0
    Win32 Version: 1.0.0.0
    CodeBase: file:///D:/PipeToCom/NP2COMV.exe
----------------------------------------
System.Windows.Forms
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.79.0 built by: NETFXREL2
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Windows.Forms/v4.0_4.0.0.0__b77a5c561934e089/System.Windows.Forms.dll
----------------------------------------
System
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.79.0 built by: NETFXREL2
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System/v4.0_4.0.0.0__b77a5c561934e089/System.dll
----------------------------------------
System.Drawing
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.79.0 built by: NETFXREL2
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Drawing/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Drawing.dll
----------------------------------------
log4net
    Assembly Version: 1.2.10.0
    Win32 Version: 1.2.10.0
    CodeBase: file:///D:/PipeToCom/log4net.DLL
----------------------------------------
System.Configuration
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.79.0 built by: NETFXREL2
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Configuration/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Configuration.dll
----------------------------------------
System.Core
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.79.0 built by: NETFXREL2
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Core/v4.0_4.0.0.0__b77a5c561934e089/System.Core.dll
----------------------------------------
System.Xml
    Assembly Version: 4.0.0.0
    Win32 Version: 4.6.79.0 built by: NETFXREL2
    CodeBase: file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Xml/v4.0_4.0.0.0__b77a5c561934e089/System.Xml.dll
----------------------------------------
NP2COM
    Assembly Version: 1.0.0.0
    Win32 Version: 1.0.0.0
    CodeBase: file:///D:/PipeToCom/NP2COM.DLL
----------------------------------------

************** JIT Debugging **************
To enable just-in-time (JIT) debugging, the .config file for this
application or computer (machine.config) must have the
jitDebugging value set in the system.windows.forms section.
The application must also be compiled with debugging
enabled.

For example:

<configuration>
    <system.windows.forms jitDebugging="true" />
</configuration>

When JIT debugging is enabled, any unhandled exception
will be sent to the JIT debugger registered on the computer
rather than be handled by this dialog box.

**Comment by sonvoba** [comment]
FWIW, I do have .NET 3.5 and earlier installed on this system.  Could it be something else?

![untitled 06d](https://cloud.githubusercontent.com/assets/10577963/9076572/ff73a696-3ad8-11e5-8112-0b03f5636239.png)

**Comment by albertjan** [comment]
@sonvoba no it's a known issue in the np2comv program. There's a namedpipe in the list that contains something I didn't expect when I wrote it.

You could edit the `.n2c` file yourself.

**Comment by sonvoba** [comment]
Albert-Jan,

Many thanks for the follow-up.  Will the np2com service know where to pick up the correct .n2c configuration file, or should I just edit the existing example.n2c and things will work?

Regards,

--svb

**Comment by albertjan** [comment]
It will pick up any n2c file in the directory. So yeah you can edit example.n2c

**Comment by sonvoba** [comment]
Thanks!

**Comment by bschappel** [comment]
How do I disable logging?

**Comment by albertjan** [comment]
@bschappel delete log4net.config

**Comment by bschappel** [comment]
Quick question regarding the serial settings in the .n2c files. Do those settings hardwire the host comport?

We've go a barcode scanner base station that connects over serial. It needs 9600-8N1. I set this in my n2c file but when I try and connect from the guests OS the software complains that there is a mismatch between the serial port and the required settings.

It may be that your software won't work with this device and that's fine.

I was just wondering if the guest OS can negotiate the settings of the comport on the host.

**Comment by albertjan** [comment]
@bschappel No it can't communicate from guest os to the np2com software. The settings for the comport in the n2c file are the ones used to connect to the real comport. The settings in the guest don't really matter. But you could be running into a CTS RTS issue. Those settings aren't comunicated to the host os. 

Otherwise it's smart to set the comport on the guest os to the same settings otherwise it'll drop bytes, because it sends data to quickly.

Remember to turn off your uart buffer on the guest os.

**Comment by kasaio** [comment]
I set & used this a few weeks back I must thank the owner\creator as after a few issues, this work very well. 
I just have one question & it's not a big - after a few weeks it seems the service becomes a little unstable & crashes out. Restarting the service is not a problem, it works again without issue until it crashes. Crashing once or twice a day & having to restart the service isn't that great. Has anyone any ideas as to what is the cause or fix to the issue?

Event log shows a .Net run-time error on each crash - 
_Application: NP2COMS.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.IO.IOException
   at System.IO.Pipes.PipeStream.WinIOError(Int32)
   at System.IO.Pipes.PipeStream.BeginWriteCore(Byte[], Int32, Int32, System.AsyncCallback, System.Object)
   at System.IO.Pipes.PipeStream.WriteCore(Byte[], Int32, Int32)
   at NP2COM.Connection.NamedPipeRunner(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Threading.ThreadHelper.ThreadStart(System.Object)_

**Comment by xiongqr** [comment]
Hi Albert,

Thanks for program.  I managed to get it to work on my XP mode VM, only one serial port though.

Does your program support multiple serial ports?  Experimented with several config files, no success.

thanks

**Comment by albertjan** [comment]
That normally does the trick several config files. If that doesn’t work. I don’t know I haven’t really done much with this for ages.


________________________________
